### PR TITLE
#9031 GetMap format list in GetCapabilities not respected

### PR DESCRIPTION
**Pull request by allyoucanmap**
## Description

This PR includes following improvements:

- refactor of WMS api removing the JSONX parser
- review how the format are included inside the layer object created from a WMS service
- review interface of getLayerFromRecord method of catalog api
- include a refresh button for the format selector of service advanced settings

### For Testers

we need to cover functional test for following components

- visibility settings of a layer
- format settings of a layer
- export of a WMC file
- style editor
- WMS/WMTS/CSW catalog
- add WMS background layer from UI (check if available styles are correctly displayed)
- layer metadata

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Refactoring

## Issue

**What is the current behavior?**

#9031
#8906

**What is the new behavior?**

The format for the WMS layer are extracted directly from the capabilities even if the advanced setting are not changed, in addition the filtering of the supported format types has been improved to allow only the supported ones listed in the capabilities

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)

 - [x] No

### CODE REVIEW

File: web/client/api/WMS.js [hunk 1]
```diff
@@ -100,11 +198,18 @@ export const searchAndPaginate = (json = {}, startPosition, maxRecords, text) =>
...
-            version: (json.WMS_Capabilities || json.WMT_MS_Capabilities)?.$?.version || '1.3.0'
+            version: getCapabilityRoot(json)?.$?.version || WMS_GET_CAPABILITIES_VERSION
         },
         records: filteredLayers
             .filter((layer, index) => index >= startPosition - 1 && index < startPosition - 1 + maxRecords)
-            .map((layer) => assign({}, layer, { formats: rootFormats, onlineResource, SRS: SRSList, credits: layer.Attribution ? extractCredits(layer.Attribution) : credits}))
+            .map((layer) => ({
+                ...layer,
+                supportedGetMapFormats,
+                supportedGetFeatureInfoFormats,
```
Comment (offtherailz) [inline web/client/api/WMS.js:208]: At the service API level we should return **all** the available formats provided by the service.
The filtering should be applied later.
```suggestion
                supportedGetMapFormats,
                supportedGetFeatureInfoFormats,
```

File: web/client/api/WMS.js [hunk 2]
```diff
@@ -84,12 +181,13 @@ export const getOnlineResource = (c) => {
     return c.Request && c.Request.GetMap && c.Request.GetMap.DCPType && c.Request.GetMap.DCPType.HTTP && c.Request.GetMap.DCPType.HTTP.Get && c.Request.GetMap.DCPType.HTTP.Get.OnlineResource && c.Request.GetMap.DCPType.HTTP.Get.OnlineResource.$ || undefined;
 };
 export const searchAndPaginate = (json = {}, startPosition, maxRecords, text) => {
```
Comment (offtherailz) [inline web/client/api/WMS.js:183]: Only one reminder: 
These functions (`searchText` and `getRecords`) are not part of WMS protocol. 
I started to isolate specific protocols APIs (WMS, WMTS...) from catalog abstraction, but we have not finished yet. 
They will need sooner or later to be moved in `api/catalog/WMS`. Anyway in this PR we can highlight the changes on the specific task. So I should move them in a separate PR. 

Try only to keep isolated the 2 concepts. (WMS protocol vs Catalog abstraction).

**Review by offtherailz — CHANGES_REQUESTED** [review]
General thing looks good, I love we are removing ogc-schemas one piece at time :heart: 

I asked for some fixes and doc/naming improvements. See my comments inline and contact me if something is unclear.

File: web/client/translations/data.it-IT.json [hunk 3]
```diff
@@ -1483,7 +1483,8 @@
             },
             "format": {
                 "noOption": "Nessuna opzione",
-                "loading": "Caricamento in corso..."
+                "loading": "Caricamento in corso...",
+                "refresh": "Recupera il formato supportato"
```
Comment (offtherailz) [inline web/client/translations/data.it-IT.json:1487]: I think it should be plural also on the other translations.
```suggestion
                "refresh": "Recupera i formati supportati"
```

**Review by offtherailz — CHANGES_REQUESTED** [review]
- [ ] Please fix the translations

The rest is ok.

**Review by offtherailz — APPROVED** [review]

**Comment by landryb** [comment]
> As exchanged in our call I would like to remove the support for _supportedGetMapFormats_ (the name can be also misunderstood) in favor of a fixed list of a well known and tested formats if we cannot find a reliable heuristic that automatically interpret it from the wms getcapa and generate that list. I wouldn't like to stratify config properties like this in favor of a more linear behavior: if we will need to include a new format we will provide a new PR to update the list of explicitly supported ones.

I'm a bit lost.. _supportedGetMapFormats_ allowed one to 'override' the hardcoded list of _formats supported by mapstore_ ? The newly hardcoded list after ea42e62f5c is a bit short.. in my opinion it should be whatever formats the browser is able to display, no ? or only the list of formats supported by OL/leaflet/cesium ? Should that only be 'official' mimetypes ?

what about `image/tiff` ? i also have servers with `image/png; mode=32bit` which .. is a valide mimetype (24-bits png with alpha channel). im not so sure too, but there might be servers out there with `image/jpg` without the e in jpeg..

I've tested `image/png; mode=8bit` from https://wms.craig.fr/ign (reported in #8906) and its now correctly available, and can be used as a default for the layers added from the service, be it in the layers display properties or the service advanced properties.

fetching available formats from https://tiles.craig.fr/ortho/service correctly only lists `image/png` & `image/jpeg` so georchestra/mapstore2-georchestra/#301 is properly handled.

will test with more services from https://geoservices.ign.fr/services-web-essentiels & https://geoservices.ign.fr/services-web-experts and others we might use.

**Comment by landryb** [comment]
https://wxs-gpu.mongeoportail.ign.fr/externe/vkd1evhid6jdj5h4hkhyzjto/wms/v?service=WMS&version=1.3.0&request=GetCapabilities provides a large list, from which `image/tiff`, `image/geotiff`, `image/tiff8` & `image/geotiff8` might appear valid in addition to the already supported ones ?

https://wxs.ign.fr/administratif/geoportail/r/wms?service=WMS&version=1.3.0&request=GetCapabilities provides:
```
<Format>image/jpeg</Format>
<Format>image/png</Format>
<Format>image/tiff</Format>
<Format>image/geotiff</Format>
<Format>image/x-bil;bits=32</Format>
<Format>text/asc</Format>
```
from this list, ms2 only displays jpeg & png. So, tiff and geotiff are somewhat popular ? are those supported by OL ?

Will do more tests tmrw, including GFI formats (eg #9007 / georchestra/mapstore2-georchestra#302)

**Comment by landryb** [comment]
for GFI formats testing, first some tests with mapserver:
- i've added `continents` layer https://demo.mapserver.org/cgi-bin/wms (mapserver 8 official demo server) which only advertises `text/html`, `application/vnd.ogc.gml`, `text/plain` - the feature info pane of the layer properties correctly lists plain and html.
  - querying with `info_format=text/plain` triggers an error:
```msWMSLoadGetMapParams(): WMS server error. Missing required parameter STYLES. Note to service administrators: defining the "wms_allow_getmap_without_styles" "true" MAP.WEB.METADATA item will disable this check (backward compatibility with behaviour of MapServer < 8.0) ```
it's the same error reported in geosolutions-it/MapStore2#8599. the STYLES parameter is also mandatory for getfeatureinfo requests. Adding an empty `STYLES=` to the query triggers another strange error:
```
msWMSFeatureInfo(): WMS server error. Layer(s) specified in QUERY_LAYERS parameter is not offered by the service instance.
```
  - this last one doesnt make sense as the query correctly specifies the same layer in both params: `layers=continents&query_layers=continents&`
- adding layers from https://wms.craig.fr/ign (mapserver 8), same issue, missing STYLES parameter. But the correct list of GetFeatureInfo formats (text/html/properties/template) is proposed in the feature info pane
- adding layers from https://wms.craig.fr/mnt (mapserver 8, this one sets `wms_allow_getmap_without_styles`):
  - the available formats are correctly listed (text/html/properties/template), i can correctly query a vector layer via html, properties or plain text. 
  - For template format, i havent been able to edit it, clicking on the pencil button doesnt show a template editor, and i have this error in the console:
```
[React Intl] Cannot format message: "layerProperties.templateFormatInfoAlertExample", using message source as fallback. 
```
this bit is another issue ?

- adding  `wms_allow_getmap_without_styles` to https://wms.craig.fr/ign and removing extra GFI formats:
  - the defaults advertised in getcapabilities are only `application/vnd.ogc.gml` and `text/plain`.
  - Loading a layer from that entrypoint, only `text` is available in the feature info pane. That's expected.

some tests with mapproxy:
- loading layers from https://tiles.craig.fr/ign (which advertises `text/plain`, `text/html`, `text/xml` & `application/json`), i can correctly query layers with the available formats (still no template working)
- removing extra formats from the mapproxy config, the defaults advertised are  `text/plain`, `text/html`, `text/xml`. As expected, only `text` and `html` are available in the feature info pane.

some tests with IGN national services:
- adding layer from https://wxs.ign.fr/administratif/geoportail/r/wms? the formats are correctly listed. querying a feature (https://wxs.ign.fr/administratif/geoportail/r/wms?service=WMS&version=1.1.1&request=GetFeatureInfo&exceptions=application/json&id=ADMINEXPRESS-COG.LATEST__19d56bf0-cd48-11ed-8463-3febeb0b92ee&layers=ADMINEXPRESS-COG.LATEST&query_layers=ADMINEXPRESS-COG.LATEST&x=51&y=51&height=101&width=101&srs=EPSG:3857&bbox=380863.019702261,5733076.897858525,381104.27407278615,5733318.15222905&feature_count=10&info_format=text/plain&ENV=mapstore_language:en) triggers a 400 code from the service:
```
<ServiceExceptionReport xmlns="http://www.opengis.net/ogc">
<ServiceException code="MissingParameterValue" >
  Parametre FORMAT absent.
</ServiceException>
</ServiceExceptionReport>
```

either FORMAT is mandatory in the protocol and should be the same value as INFO_FORMAT, or IGN service is too picky/broken.

and a last test with the GPU national service:
- adding `du` layer from https://wxs-gpu.mongeoportail.ign.fr/externe/vkd1evhid6jdj5h4hkhyzjto/wms/v, the format list is correct
- i can correctly query layers using plain, html and properties formats. Great! The template format still yields the error.


I hope this testing covers as much cases as possible. To me there are two issues to worry about:
- the missing mandatory `STYLES` parameter, i'll eventually get around tackling it
- the broken templates format, unsure if its been broken by this PR or something else.

**Comment by allyoucanmap** [comment]
> I'm a bit lost.. supportedGetMapFormats allowed one to 'override' the hardcoded list of formats supported by mapstore ? The newly hardcoded list after https://github.com/geosolutions-it/MapStore2/commit/ea42e62f5c60fe4b473b97a77719b90ca51c6ca5 is a bit short.. in my opinion it should be whatever formats the browser is able to display, no ? or only the list of formats supported by OL/leaflet/cesium ? Should that only be 'official' mimetypes ?

> what about image/tiff ? i also have servers with image/png; mode=32bit which .. is a valide mimetype (24-bits png with alpha channel). im not so sure too, but there might be servers out there with image/jpg without the e in jpeg..

@landryb We removed the proposed implementation of `supportedGetMapFormats`  because every new format we will include in the list must be tested and we need to ensure the browser and map libraries can properly handle it. So we could add the `image/png; mode=32bit` mime type to the list if you have a server that could be tested

> https://wxs-gpu.mongeoportail.ign.fr/externe/vkd1evhid6jdj5h4hkhyzjto/wms/v?service=WMS&version=1.3.0&request=GetCapabilities provides a large list, from which `image/tiff`, `image/geotiff`, `image/tiff8` & `image/geotiff8` might appear valid in addition to the already supported ones ?
> 
> https://wxs.ign.fr/administratif/geoportail/r/wms?service=WMS&version=1.3.0&request=GetCapabilities provides:
> 
> ```
> <Format>image/jpeg</Format>
> <Format>image/png</Format>
> <Format>image/tiff</Format>
> <Format>image/geotiff</Format>
> <Format>image/x-bil;bits=32</Format>
> <Format>text/asc</Format>
> ```
> 
> from this list, ms2 only displays jpeg & png. So, tiff and geotiff are somewhat popular ? are those supported by OL ?

We tested tiff, tiff8, geotiff and geotiff8 from a GeoServer but they are not rendered by the browser and OL. The map library is using the [Image](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/Image) class that is equivalent to the `` tag and the most [common supported formats](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#supported_image_formats) do not include tiff nor geotiff.
The default list of supported format in MapStore could include also "unofficial" mime type but we need at least a server with the implementation that could be properly tested, so if you have one with `image/jpg` we could add also this to the list, thanks.
In general the idea is this https://github.com/geosolutions-it/MapStore2/pull/9054#pullrequestreview-1359014176:

>  if we will need to include a new format we will provide a new PR to update the list of explicitly supported ones.

**Review by tdipisa — APPROVED** [review]
Thank you @allyoucanmap and @offtherailz for the review.
@landryb, according to https://github.com/geosolutions-it/MapStore2/pull/9054#issuecomment-1488097252 if you have further testable formats to add to the list, just let us know and we will raise a new PR for them.

@allyoucanmap, I think the only things that are still missing are:
- Provide a comment in #9007 with further indications for @landryb 
- Provide all missing backports

**Comment by tdipisa** [comment]
@ElenaGallo please test this in DEV following indications provided by @allyoucanmap in PR description. No backport needed for this since it is for 2023.02.00

**Comment by landryb** [comment]
> So we could add the `image/png; mode=32bit` mime type to the list if you have a server that could be tested

sorry, that server isnt public.

> We tested tiff, tiff8, geotiff and geotiff8 from a GeoServer but they are not rendered by the browser and OL. The map library is using the [Image](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/Image) class that is equivalent to the `` tag and the most [common supported formats](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#supported_image_formats) do not include tiff nor geotiff.

Right, then it makes sense to not add them. Thanks for the explanation.

**Pull request merged**